File: utils/src/main/java/com/github2136/util/JsonUtil.java
package com.github2136.util;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;

import java.lang.reflect.Type;

/**
 * Json工具类JsonUtilCompat</br>
 * 如果需要使用其他的GsonBuilder，继承该类重写<b>getInstance</b>/<b>getGson</b>
 */
public class JsonUtil {
    private Gson mGson;

    public static JsonUtil getInstance() {
        return new JsonUtil();
    }

    protected JsonUtil() {
        mGson = new GsonBuilder().setDateFormat("yyyy-MM-dd HH:mm:ss").create();
    }

    public Gson getGson() {
        return mGson;
    }

    public <T> T getObjectByStr(String json, Class<T> cls) {
        try {
            return mGson.fromJson(json, cls);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public <T> T getObjectByStr(String json, Type typeOfT) {
        try {
            return mGson.fromJson(json, typeOfT);
        } catch (JsonSyntaxException e) {
            e.printStackTrace();
            return null;
        }
    }

    public String toJsonStr(Object src) {
        return mGson.toJson(src);
    }
}

File: utils/src/main/java/com/github2136/util/JsonUtil.kt
package com.github2136.util

import android.util.Log
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import java.io.PrintWriter
import java.io.StringWriter

/**
 * Created by yb on 2018/8/24.
 */
class JsonUtil private constructor(val gson: Gson) {

    inline fun <reified T> fromJson(json: String?): T? {
        return try {
            gson.fromJson<T>(json, object : TypeToken<T>() {}.type)
        } catch (e: JsonSyntaxException) {
            val stringWriter = StringWriter()
            e.printStackTrace(PrintWriter(stringWriter))
            Log.e("JsonUtil", stringWriter.toString())
            null
        }
    }

    fun toJson(obj: Any?): String {
        return gson.toJson(obj)
    }

    companion object {
        var dateFormat = DateUtil.DATE_PATTERN_YMDHMS
        var mGson = GsonBuilder().setDateFormat(dateFormat).create()

        val instance: JsonUtil by lazy {
            JsonUtil(mGson)
        }

        fun newInstance(gson: Gson = GsonBuilder().setDateFormat(DateUtil.DATE_PATTERN_YMDHMS).create()): JsonUtil {
            return JsonUtil(gson)
        }
    }
}